File: get_HMM_domains.pl
#!/usr/bin/env perl

#Wrapper script to run Hmmer3 to find coordinates of domain, reformat, and write just the domain for each to new file

use strict;
use warnings;
use Getopt::Long;

#set command line arguments
my ($proteins, $hmm, $length) = @ARGV;
my $version="get_HMM_domains.pl\tv0.1.1";
my $osname = $^O;
my $helpAsked;
my $hmm_db="/Users/jon/projects/DB/HMM"; #need to set this to your local folder containing HMM models
my $cpus = 1;
my $evalue = 1e-10;
my $outfile = "default_HMM_output.fasta";
GetOptions(
    'p|proteins:s'=>\$proteins, #protein fasta file
    'm|hmm:s'=>\$hmm, #name of HMM model to search
    'e|evalue:s'=>\$evalue, #directory you are working in
    'l|length:s'=>\$length, #min length required to keep domain
    'c|cpus:i'=>\$cpus, #cpus to use for hmmer search, default is one
    'o|out:s'=>\$outfile, # final multi-fasta file
    'h|help:s'=>\$helpAsked,
    'v|version'=>sub{print $version."\n"; exit;},
);

if(defined($helpAsked)) {
    prtUsage();
    exit;
}
if(!defined($proteins)) {
    print "You did not specify a protein fasta file, type get_HMM_domains.pl -h for help\n";
    exit;
}
print "-------------------------------------------------------------\n";
#check to see if required external programs are in path
my $check_hmm = `hmmscan -h | grep "# HMMER" | sed "s/# //" `;
my $check_bed = `bedtools --version`;

if (defined($check_hmm) && $check_hmm ne "") {
  print "\nFound HMMer: $check_hmm";
}
else {
  print "HMMer installation not found in PATH, exiting\n";
  exit;
}
if (defined($check_bed) && $check_bed ne "") {
  print "Found Bedtools: $check_bed\n";
}
else {
  print "BedTools installation not found in PATH, exiting\n";
  exit;
}
#Sub routine for running hmmscan and creating a bed file
sub runHMMer{
print "Running hmmerscan:\n   hmmscan --domtblout hmmscan.temp.out --cpu $cpus -E $evalue $hmm_db/$hmm $proteins\n";
my $hmm_results = `hmmscan --domtblout hmmscan.temp.out --cpu $cpus -E $evalue $hmm_db/$hmm $proteins`;
}
sub formatHMM{
my $output = 'hmmscan_domains.bed';
open(OUTPUT, '>'.$output) or die;
open(my $fh, "<:encoding(UTF-8)", 'hmmscan.temp.out') or die;
open(OUTPUT, '>>'.$output) or die;
while (<$fh>) {
    next if /^#/;
    my @row = split(/\s+/,$_);
    my $len = $row[20]-$row[19];
    if ($len > $length) {
        print OUTPUT "$row[3]\t$row[19]\t$row[20]\n";
    }
}
close $fh;
close $output;
}

sub getSeqs{
print "\nGetting Fasta coordinates:\n   bedtools getfasta -fullHeader -fi $proteins -bed hmmscan_domains.bed -fo $outfile\n";
my $fasta_seqs = `bedtools getfasta -fullHeader -fi $proteins -bed hmmscan_domains.bed -fo $outfile`
}

#sub routine for help
sub prtHelp {
    print "\n";
    print "All Options are Required:\n";
    print "-------------------------------------------------------------\n";
    print "--proteins flag:  fasta protein file \n";
    print "--hmm flag:  HMM model to use (KS or A)\n";
    print "--out flag:  Output file \n";
    print "--length flag:  Minimum length of domain\n";
    print "--evalue flag:  Evalue cutoff to use \n";
    print "--cpu flag:  Number of threads to use\n";
    print "-------------------------------------------------------------\n";
    print "example:  get_HMM_domains.pl --proteins proteins.fasta --hmm KS.hmm --length 300 --out domains_out.fasta 1\n";
}
#print usage subroutine
sub prtUsage {
    print "OS version: $osname\n";
    print "Usage: perl $0 -h\n";
    prtHelp();
}

#Run sub routines
runHMMer();
formatHMM();
getSeqs();


print "\n";
print "Script has successfully run, your $hmm domain proteins are in $outfile\n";
print "-------------------------------------------------------------\n";
unlink 'hmmscan_domains.bed';
unlink 'hmmscan.temp.out';

File: getHMMdomains.py
#!/usr/bin/env python

#script takes input protein fasta and HMM model to pull out those domains
#written by Jon Palmer palmer.jona at gmail dot com
import warnings
with warnings.catch_warnings():
    warnings.simplefilter('ignore')
    from Bio import SearchIO
import sys
import os
from Bio import SeqIO
import argparse
from os.path import expanduser
home = expanduser("~")

#define HMM db here
hmm_db = home + '/projects/DB/HMM'

class MyFormatter(argparse.ArgumentDefaultsHelpFormatter):
    def __init__(self,prog):
        super(MyFormatter,self).__init__(prog,max_help_position=50)
parser=argparse.ArgumentParser(prog='getHMMdomains.py', usage="%(prog)s [options] proteins.fasta > domains.fasta\n%(prog)s -h for help menu",
    description='''Script to identify HMM domains from proteins and pull out only sequence corresponding to the domain in multi-fasta format.  HMMER3 is a dependency''',
    formatter_class=MyFormatter,
    epilog="""Written by Jon Palmer (2015) palmer.jona@gmail.com""")
parser.add_argument('fasta', help='Fasta input (aa seq)')
parser.add_argument('-m','--hmm', default='KS_domain.hmm', help='HMM model (Can also specificy full path)')
parser.add_argument('-e','--evalue', default="1e-10", help='HMM Evalue cutoff')
parser.add_argument('-l','--length', default="50", help='min length of domain to keep')
parser.add_argument('-c','--cpus', default="1", help='HMM model')
args=parser.parse_args()
    
fasta = args.fasta
if not "/" in args.hmm:
    hmm = hmm_db + '/' + args.hmm
else:
    hmm = args.hmm
evalue = args.evalue
length = args.length
cpus = args.cpus

#run hmmerscan search
os.system("%s %s %s %s %s %s %s" % ('hmmscan --cpu', cpus, '-E', evalue, hmm, fasta, '> hmmscan.temp.txt'))

#format results and print fasta seq based on coordinates
hmmer_results = open('hmmscan.temp.txt', 'r')
inputSeqFile = open(fasta, "rU")
SeqRecords = SeqIO.to_dict(SeqIO.parse(inputSeqFile, "fasta"))
for qresult in SearchIO.parse(hmmer_results, "hmmer3-text"):
    hits = qresult.hits
    if len(hits) > 0:
        beste = hits[0].hsps[0].evalue
        query = hits[0].query_id
        hit = hits[0].id.replace('.hmm', '')
        start = hits[0].hsps[0].env_start
        end = hits[0].hsps[0].env_end
        hit_length = end - start
        if int(hit_length) >= int(length) and float(beste) <= float(evalue):
            subseq = SeqRecords[query][start:end].seq
            print("%s%s|%s|%s-%s\n%s" % ('>', query, hit, start, end, subseq))

#close and clean
hmmer_results.close()
inputSeqFile.close()
os.remove('hmmscan.temp.txt')